src/Integration.Vsix/ErrorList/SonarErrorListDataSource.cs

```diff
@@ -0,0 +1,139 @@
+{
+    // Interface introduced to simplify testing.
+    internal interface ISinkManagerRegister
+    {
+        void AddSinkManager(ISinkManager manager);
+        void RemoveSinkManager(ISinkManager manager);
+    }
+
+    [Export(typeof(ISonarErrorListDataSource))]
+    internal class SonarErrorListDataSource : ITableDataSource, ISonarErrorListDataSource, ISinkManagerRegister
+    {
+        private readonly ISet<ISinkManager> managers = new HashSet<ISinkManager>();
+        private readonly ISet<ITableEntriesSnapshotFactory> factories = new HashSet<ITableEntriesSnapshotFactory>();
+
+        [ImportingConstructor]
+        internal SonarErrorListDataSource(ITableManagerProvider tableManagerProvider)
+        {
+            if (tableManagerProvider == null)
+            {
+                throw new ArgumentNullException(nameof(tableManagerProvider));
+            }
+
+            var errorTableManager = tableManagerProvider.GetTableManager(StandardTables.ErrorsTable);
+            errorTableManager.AddSource(this, StandardTableColumnDefinitions.DetailsExpander,
+                                                   StandardTableColumnDefinitions.ErrorSeverity, StandardTableColumnDefinitions.ErrorCode,
+                                                   StandardTableColumnDefinitions.ErrorSource, StandardTableColumnDefinitions.BuildTool,
+                                                   StandardTableColumnDefinitions.ErrorSource, StandardTableColumnDefinitions.ErrorCategory,
+                                                   StandardTableColumnDefinitions.Text, StandardTableColumnDefinitions.DocumentName,
+                                                   StandardTableColumnDefinitions.Line, StandardTableColumnDefinitions.Column,
+                                                   StandardTableColumnDefinitions.ProjectName);
+        }
+
+        #region ITableDataSource members
+
+        public string DisplayName => "SonarLint";
+
+        public string Identifier => "SonarLint";
+
+        public string SourceTypeIdentifier => StandardTableDataSources.ErrorTableDataSource;
+
+        // Note: Error List is the only expected subscriber
+        public IDisposable Subscribe(ITableDataSink sink) => new SinkManager(this, sink);
+
+        #endregion
+
+        #region ISinkManagerRegister
+
+        public void AddSinkManager(ISinkManager manager)
+        {
+            lock (managers)
+            {
+                managers.Add(manager);
+
+                foreach (var factory in factories)
+                {
+                    manager.AddFactory(factory);
+                }
+            }
+        }
+
+        public void RemoveSinkManager(ISinkManager manager)
+        {
+            lock (managers)
+            {
+                managers.Remove(manager);
+            }
+        }
+
+        #endregion
+
+        #region ISonarErrorListDataSource implementation
+
+        public void RefreshErrorList()
```

Comment:
This method tells the error list to update the issues for all of the files we've analysed. We should be able to tell the error list which factory (i.e. file with issues) has changed. But I didn't want to change that as part of this refactoring.